pkg/storage/azure/azure.go

```diff
@@ -0,0 +1,506 @@
+	}
+
+	service := azblob.NewServiceURL(*u, p)
+	return service.NewContainerURL(containerName), nil
+}
+
+func createStorageContainer(accountName, key, containerName string) error {
+	ctx := context.TODO()
+
+	container, err := getStorageContainer(accountName, key, containerName)
+	if err != nil {
+		return err
+	}
+
+	_, err = container.Create(ctx, azblob.Metadata{}, azblob.PublicAccessNone)
+	return err
+}
+
+func deleteStorageContainer(accountName, key, containerName string) error {
+	ctx := context.TODO()
+
+	container, err := getStorageContainer(accountName, key, containerName)
+	if err != nil {
+		return err
+	}
+
+	_, err = container.Delete(ctx, azblob.ContainerAccessConditions{})
+	return err
+}
+
+type driver struct {
+	Config     *imageregistryv1.ImageRegistryConfigStorageAzure
+	KubeConfig *rest.Config
+	Listers    *regopclient.Listers
+}
+
+// NewDriver creates a new storage driver for Azure Blob Storage.
+func NewDriver(c *imageregistryv1.ImageRegistryConfigStorageAzure, kubeconfig *rest.Config, listers *regopclient.Listers) *driver {
+	return &driver{
+		Config:     c,
+		KubeConfig: kubeconfig,
+		Listers:    listers,
+	}
+}
+
+func (d *driver) getAzureConfig() (*clusterconfig.Azure, error) {
+	cfg, err := clusterconfig.GetAzureConfig(d.Listers)
+	if err != nil {
+		return nil, err
+	}
+	return cfg, nil
+}
+
+func (d *driver) storageAccountsClient(cfg *clusterconfig.Azure) (storage.AccountsClient, error) {
+	auth, err := auth.NewClientCredentialsConfig(cfg.ClientID, cfg.ClientSecret, cfg.TenantID).Authorizer()
+	if err != nil {
+		return storage.AccountsClient{}, err
+	}
+
+	storageAccountsClient := storage.NewAccountsClient(cfg.SubscriptionID)
+	storageAccountsClient.Authorizer = auth
+	storageAccountsClient.AddToUserAgent(parameters.UserAgent)
+
+	return storageAccountsClient, nil
+}
+
+// ConfigEnv configures the environment variables that will be used in the
+// image registry deployment.
+func (d *driver) ConfigEnv() (envs []corev1.EnvVar, err error) {
+	envs = append(envs,
+		corev1.EnvVar{Name: "REGISTRY_STORAGE", Value: "azure"},
+		corev1.EnvVar{Name: "REGISTRY_STORAGE_AZURE_CONTAINER", Value: d.Config.Container},
+		corev1.EnvVar{Name: "REGISTRY_STORAGE_AZURE_ACCOUNTNAME", Value: d.Config.AccountName},
+		corev1.EnvVar{
+			Name: "REGISTRY_STORAGE_AZURE_ACCOUNTKEY",
+			ValueFrom: &corev1.EnvVarSource{
+				SecretKeyRef: &corev1.SecretKeySelector{
+					LocalObjectReference: corev1.LocalObjectReference{
+						Name: imageregistryv1.ImageRegistryPrivateConfiguration,
+					},
+					Key: "REGISTRY_STORAGE_AZURE_ACCOUNTKEY",
+				},
+			},
+		},
+	)
+	return
+}
+
+func (d *driver) Volumes() ([]corev1.Volume, []corev1.VolumeMount, error) {
+	return nil, nil, nil
+}
+
+// Secrets returns a map of the storage access secrets.
+func (d *driver) Secrets() (map[string]string, error) {
+	cfg, err := d.getAzureConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	key := cfg.AccountKey
+	if !cfg.UPI {
```

Comment:
Based on my reading it looks like `cfg.AccountKey` is something that is only provided by the cluster admin. Recall that even with an IPI install, a cluster admin can override operator settings.

Maybe clarify by renaming to `cfg.UserAccountKey` or `cfg.StorageAccountKey`